nni/__main__.py

```diff
@@ -29,82 +29,50 @@ def main():
-        _run_advisor(exp_params)
-    else:
-        # tuner (and assessor) is enabled and starts to run
-        assert exp_params.get('tuner') is not None
-        tuner = _create_tuner(exp_params)
+    assert exp_params.get('tuner') is not None
+    tuner = _create_algo(exp_params['tuner'], 'tuner')
+
+    if isinstance(tuner, MsgDispatcherBase):  # is advisor
+        logger.debug(f'Tuner {type(tuner).__name__} is advisor.')
         if exp_params.get('assessor') is not None:
-            assessor = _create_assessor(exp_params)
-        else:
-            assessor = None
-        dispatcher = MsgDispatcher(tuner, assessor)
-
-        try:
-            dispatcher.run()
-            tuner._on_exit()
-            if assessor is not None:
-                assessor._on_exit()
-        except Exception as exception:
-            logger.exception(exception)
-            tuner._on_error()
-            if assessor is not None:
-                assessor._on_error()
-            raise
-
-
-def _run_advisor(exp_params):
-    if exp_params.get('advisor').get('name'):
-        dispatcher = create_builtin_class_instance(
-            exp_params['advisor']['name'],
-            exp_params['advisor'].get('classArgs'),
-            'advisors')
+            logger.error('Tuner {type(tuner).__name__} does not support early stopping. Assessor is ignored.')
+        tuner.run()
+        return
+
+    if exp_params.get('assessor') is not None:
+        assessor = _create_algo(exp_params['assessor'], 'assessor')
     else:
-        dispatcher = create_customized_class_instance(exp_params.get('advisor'))
-    if dispatcher is None:
-        raise AssertionError('Failed to create Advisor instance')
+        assessor = None
+    dispatcher = MsgDispatcher(tuner, assessor)
+
     try:
         dispatcher.run()
-    except Exception as exception:
-        logger.exception(exception)
+        tuner._on_exit()
+        if assessor is not None:
+            assessor._on_exit()
+    except Exception:
+        tuner._on_error()
+        if assessor is not None:
+            assessor._on_error()
         raise
 
 
-def _create_tuner(exp_params):
-    if exp_params['tuner'].get('name'):
-        tuner = create_builtin_class_instance(
-            exp_params['tuner']['name'],
-            exp_params['tuner'].get('classArgs'),
-            'tuners')
-    else:
-        tuner = create_customized_class_instance(exp_params['tuner'])
-    if tuner is None:
-        raise AssertionError('Failed to create Tuner instance')
-    return tuner
-
-
-def _create_assessor(exp_params):
-    if exp_params['assessor'].get('name'):
-        assessor = create_builtin_class_instance(
-            exp_params['assessor']['name'],
-            exp_params['assessor'].get('classArgs'),
-            'assessors')
+def _create_algo(algo_config, algo_type):
+    if algo_config.get('name'):
+        algo = create_builtin_class_instance(algo_config['name'], algo_config.get('classArgs'), algo_type + 's')
```

Comment:
`+ 's'` is ok, but a little strange...

Reply:
I totally agree. This is historical burden.
In experiment config they are called "tuner" and "assessor", but in algorithm manifest the keys are "tuners" and "assessors".
We don't want to change the file format until 3.0.